src/go/rpk/pkg/cli/cluster/config/set.go

```diff
@@ -60,92 +77,149 @@ You may also use <key>=<value> notation for setting configuration properties:
-				// about which configuration properties were invalid.
-				if he.Response.StatusCode == 400 {
-					ve, err := formatValidationError(err, he)
-					out.MaybeDie(err, "error setting config: %v", err)
-					out.Die("No changes were made: %v", ve)
+				status, err := client.ClusterConfigStatus(cmd.Context(), true)
+				out.MaybeDie(err, "unable to check if the cluster needs to be restarted: %v\nCheck the status with 'rpk cluster config status'.", err)
+				for _, value := range status {
+					if value.Restart {
+						fmt.Print("\nCluster needs to be restarted. See more details with 'rpk cluster config status'.\n")
+						break
+					}
 				}
 			}
+		},
+	}
+
+	return cmd
+}
+
+func validateConfigSelfHosted(schema rpadmin.ConfigSchema, args []string) (map[string]interface{}, []string, error) {
+	upsert := make(map[string]interface{})
+	remove := make([]string, 0)
 
-			out.MaybeDie(err, "error setting property: %v", err)
-			fmt.Printf("Successfully updated configuration. New configuration version is %d.\n", result.ConfigVersion)
+	for _, arg := range args {
+		split := strings.SplitN(arg, "=", 2)
+		key, value := split[0], split[1]
 
-			status, err := client.ClusterConfigStatus(cmd.Context(), true)
-			out.MaybeDie(err, "unable to check if the cluster needs to be restarted: %v\nCheck the status with 'rpk cluster config status'.", err)
-			for _, value := range status {
-				if value.Restart {
-					fmt.Print("\nCluster needs to be restarted. See more details with 'rpk cluster config status'.\n")
+		meta, ok := schema[key]
+
+		if !ok {
+			// loop over schema, try to find key in the Aliases,
+			for _, v := range schema {
+				if slices.Contains(v.Aliases, key) {
+					meta, ok = v, true
 					break
 				}
 			}
+			if !ok {
+				return nil, nil, fmt.Errorf("Unknown property %q", key)
+			}
+		}
+
+		// - For scalars, pass string values through to the REST
+		// API -- it will give more informative errors than we can
+		// about validation.  Special case strings for nullable
+		// properties ('null') and for resetting to default ('')
+		// - For arrays, make an effort: otherwise the REST API
+		// may interpret a scalar string as a list of length 1
+		// (via one_or_many_property).
+
+		if meta.Nullable && value == "null" {
+			// Nullable types may be explicitly set to null
+			upsert[key] = nil
+		} else if meta.Type != "string" && (value == "") {
+			// Non-string types that receive an empty string
+			// are reset to default
+			remove = append(remove, key)
+		} else if meta.Type == "array" {
+			var a anySlice
+			err := yaml.Unmarshal([]byte(value), &a)
+			if err != nil {
+				return nil, nil, fmt.Errorf("invalid list syntax: %v", err)
+			}
+			upsert[key] = a
+		} else {
+			upsert[key] = value
+		}
+	}
+
+	return upsert, remove, nil
+}
+
+func setCloudConfig(cmd *cobra.Command, cfg *config.Config, p *config.RpkProfile, configs []string) (*controlplanev1.UpdateClusterOperation, error) {
```

Comment:
The signature only uses `cmd` to get the context, instead, why don't you accept a `context.Context` in the signature and pass a context from the caller?

Reply:
make a lot sense